[R4R]: Vesting Account(s) Implementation

alexanderbez opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Implementation of the vesting account specification (per #2589):

- Updated and cleaned up specification. As implementation went along, I found some structural mistakes and found room to DRY/clean things up.
- Implemented `sdk.MaxInt` and `sdk.MaxUint`.
- Implemented continuous and delayed (discrete) vesting account types

----

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [x] Updated relevant documentation (`docs/`)
- [x] Added entries in `PENDING.md` with issue # 
- [x] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

fedekunze commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+}
+
+//-----------------------------------------------------------------------------
+// Codec
```
shouldn't we split this file into the different account types + other (eg: `codec`)

alexanderbez replied:
Perhaps. The definitions are short, but I'm indifferent overall 👍

alexanderbez replied:
Actually, I think this might be overkill as the definitions are pretty succinct.

fedekunze commented on x/bank/keeper.go:
```diff
@@ -258,3 +297,56 @@ func inputOutputCoins(ctx sdk.Context, am auth.AccountKeeper, inputs []Input, ou
+		}
+	}
+
+	return sdk.NewTags("sender", []byte(addr.String())), nil
```
shouldn't we add an `action` here as well ?

alexanderbez replied:
Good question. Is this the standard now?

fedekunze replied:
I guess so... Maybe @cwgoes or @sunnya97 will know better?

alexanderbez replied:
Done.

fedekunze commented on x/bank/keeper.go:
```diff
@@ -258,3 +297,56 @@ func inputOutputCoins(ctx sdk.Context, am auth.AccountKeeper, inputs []Input, ou
+		}
+	}
+
+	return sdk.NewTags("recipient", []byte(addr.String())), nil
```
ditto

alexanderbez replied:
Updated.

alexanderbez commented on docs/spec/auth/vesting.md:
```diff
@@ -126,100 +120,85 @@ Thus, the total amount of _vested_ coins is `V'` and the remaining amount, `V`,
 
 #### Delayed/Discrete Vesting Accounts
 
 Delayed vesting accounts are easier to reason about as they only have the full
```
Yes, updated.

alexanderbez commented on x/stake/keeper/delegation.go:
```diff
@@ -374,10 +374,9 @@ func (k Keeper) Delegate(ctx sdk.Context, delAddr sdk.AccAddress, bondAmt sdk.Co
+		_, err := k.bankKeeper.DelegateCoins(ctx, delegation.DelegatorAddr, sdk.Coins{bondAmt})
 		if err != nil {
-			return
+			return newShares, err
```
To have explicit return variables. `newShares` is the default value for `sdk.Dec` here. I can alternatively return `sdk.Dec{}` here if you'd like.

alexanderbez replied:
Actually, I think that's more clear anyway. So I'll update

cwgoes replied:
Did you update?

alexanderbez replied:
Hmmm I swear I did. Updated again.

alexanderbez commented on x/bank/keeper.go:
```diff
@@ -258,3 +297,56 @@ func inputOutputCoins(ctx sdk.Context, am auth.AccountKeeper, inputs []Input, ou
+
+	ctx.GasMeter().ConsumeGas(costSubtractCoins, "delegateCoins")
+
+	oldCoins := getCoins(ctx, ak, addr)
```
Which logic exactly @cwgoes? There isn't that much overlap between `undelegateCoins` and `delegateCoins`?

cwgoes replied:
There's some (we just switch `Minus` and `Plus`, and switch `TrackDelegation` for `TrackUndelegation`). Not a big deal though.

alexanderbez commented on x/auth/ante.go:
```diff
@@ -256,23 +257,33 @@ func adjustFeesByGas(fees sdk.Coins, gas int64) sdk.Coins {
+// updated account and a result is returned.
+func deductFees(ctx sdk.Context, acc Account, fee sdk.Coins) (Account, sdk.Result) {
+	// TODO: Do we need charge gas for fee deduction?
+	oldCoins := acc.GetCoins()
```
Hmm so I can make a `SubtractCoinsAndSave` *auxiliary function*  in `account.go` that performs this. However, it seems it will only be used in the ante handler (as of now) even though the logic is nearly identical in `subtractCoins` in the bank keeper. The reason is the bank keeper is slightly different in that it charges gas for many ops and uses the acc keeper under the hood (abstracted away) with no direct access to an account.

Do you have any exact suggestions?

alexanderbez replied:
Actually, I may know a way of doing this.

cwgoes replied:
I see, you're right that as is it would not be possible to use the exact same function in both places. 

Mostly, I think it would be less easy to misuse in callers if we could write this logic - and the logic for `DelegateCoins` and `UndelegateCoins` - in a way that *didn't* require type-switching on `VestingAccount` but was instead a method on the account interface (so the vesting account could do one thing and the base account another). Do you think that would be possible?

alexanderbez replied:
Ok, updated. I essentially moved a few methods to the parent `Account` interface and this allowed me to cleanup and remove type checking in the bank keeper.

alexanderbez commented on x/auth/account_test.go:
```diff
@@ -106,3 +112,326 @@ func TestBaseAccountMarshal(t *testing.T) {
+	require.Nil(t, dva.DelegatedFree)
+	require.Nil(t, dva.GetCoins())
+
+	// require no modifications when delegation amount is zero or not enough funds
```
Agreed. Updated.

alexanderbez commented on x/auth/account_test.go:
```diff
@@ -106,3 +112,326 @@ func TestBaseAccountMarshal(t *testing.T) {
+	require.Nil(t, dva.DelegatedVesting)
+	require.Equal(t, origCoins, dva.GetCoins())
+
+	// require no modifications when the undelegation amount is zero
```
Agreed. Updated.

alexanderbez commented on x/bank/keeper.go:
```diff
@@ -193,33 +211,54 @@ func hasCoins(ctx sdk.Context, am auth.AccountKeeper, addr sdk.AccAddress, amt s
 		return amt, nil, sdk.ErrInsufficientCoins(fmt.Sprintf("%s < %s", oldCoins, amt))
 	}
-	err := setCoins(ctx, am, addr, newCoins)
+
```
See comment above.

cwgoes replied:
See reply to comment.

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+// GetVestedCoins returns the total amount of vested coins for a delayed vesting
+// account. All coins are only vested once the schedule has elapsed.
+func (dva DelayedVestingAccount) GetVestedCoins(blockTime time.Time) sdk.Coins {
+	if blockTime.Unix() >= dva.EndTime.Unix() {
```
You mean: `blockTime.After(dva.EndTime) || blockTime.Equal(dva.EndTime)`?

cwgoes replied:
Oh, if there isn't a `AtOrAfter` function I agree with you.

alexanderbez commented on x/auth/account_test.go:
```diff
@@ -106,3 +112,326 @@ func TestBaseAccountMarshal(t *testing.T) {
+	require.Equal(t, sdk.Coins{sdk.NewInt64Coin(testDenom, 50)}, cva.DelegatedFree)
+	require.Nil(t, cva.GetCoins())
+
+	// require no modifications when delegation amount is zero or not enough funds
```
Agreed. Updated.

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+	// We must handle the case where the start time for a vesting account has
+	// been set into the future or when the start of the chain is not exactly
+	// known.
+	if blockTime.Unix() <= cva.StartTime.Unix() {
```
Same as response below.

cwgoes replied:
See response below.

alexanderbez commented on x/auth/account_test.go:
```diff
@@ -106,3 +112,326 @@ func TestBaseAccountMarshal(t *testing.T) {
+	require.Equal(t, origCoins, cva.DelegatedFree)
+	require.Nil(t, cva.GetCoins())
+
+	// require the ability to delegate all vesting coins (50%) and all vested coins (50%)
```
Good catch. Done!

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+	}
+
+	// calculate the vesting scalar
+	x := blockTime.Unix() - cva.StartTime.Unix()
```
Updated. I have to cast to `int64`, but I don't see any immediate problems due to this.

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+	for _, ovc := range cva.OriginalVesting {
+		vestedAmt := sdk.NewDecFromInt(ovc.Amount).Mul(s).RoundInt()
+		vestedCoin := sdk.NewCoin(ovc.Denom, vestedAmt)
+		vestedCoins = vestedCoins.Plus(sdk.Coins{vestedCoin})
```
Good catch. Updated.

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+
+	bc := bva.GetCoins()
+
+	for _, coin := range bc {
```
@cwgoes I've updated this to be `O(n)`. Let me know your thoughts. It requires all the relevant coins to be sorted.

cwgoes replied:
Looks good, can we separate out that logic (zipping together two coin arrays and calling a function on the equal-denom amounts) into a helper function on the coins type for clarity?

alexanderbez replied:
@cwgoes I'm having a bit of difficulty imagining what this looks like, given in this case we're lining up one main coin set with two others.

I've gone ahead and updated `TrackUndelegation` and `trackDelegation` to use similar performance improvements. `trackDelegation` lines up one coin set with three others.

cwgoes replied:
OK, it's fine for now.

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+	for _, coin := range amount {
+		// Skip if the delegation amount is zero or if the base coins does not
+		// exceed the desired delegation amount.
+		if coin.Amount.IsZero() || bc.AmountOf(coin.Denom).LT(coin.Amount) {
```
Updated to panic. Related to your other comments.

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+	bc := bva.GetCoins()
+
+	for _, coin := range amount {
+		// skip if the undelegation amount is zero
```
Panicing now.

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+		}
+
+		// XXX: this should most likely be handled upstream (i.e. bank keeper)
+		bva.Coins = bc.Minus(sdk.Coins{coin})
```
Can you elaborate @cwgoes? This is in `x/auth`.

cwgoes replied:
I'm wondering whether we should set the account's coins here or upstream when we call this function (in accordance with the comment). It's OK here, although it's worth noting that the account isn't ever saved, so the caller must do that.

alexanderbez replied:
Hmm thinking back to when I wrote this (`XXX`), it doesn't make much sense to me anymore. Leaving the code as-is is fine imho. We're already mutating various fields of the account. In addition, we'd have to return coins here for the caller to set them...which seems redundant. So i'll leave as-is unless you have objections.

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+
+		if !y.IsZero() {
+			yCoin := sdk.NewCoin(coin.Denom, y)
+			bva.DelegatedFree = bva.DelegatedFree.Plus(sdk.Coins{yCoin})
```
Updated to be `O(n)` (see: `Coins#AddCoinByDenom` and `Coins#SubCoinByDenom`).

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+
+		if !x.IsZero() {
+			xCoin := sdk.NewCoin(coin.Denom, x)
+			bva.DelegatedFree = bva.DelegatedFree.Minus(sdk.Coins{xCoin})
```
Updated to be `O(n)` (see: `Coins#AddCoinByDenom` and `Coins#SubCoinByDenom`).

alexanderbez commented on x/auth/account.go:
```diff
@@ -121,12 +137,260 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+
+		if !y.IsZero() {
+			yCoin := sdk.NewCoin(coin.Denom, y)
+			bva.DelegatedVesting = bva.DelegatedVesting.Minus(sdk.Coins{yCoin})
```
Updated to be `O(n)` (see: `Coins#AddCoinByDenom` and `Coins#SubCoinByDenom`).

alexanderbez commented:
@cwgoes I addressed a bulk of your comments. I think a few things still need to be followed up on and so we can move from there 👍

cwgoes commented on docs/spec/auth/vesting.md:
```diff
@@ -1,34 +1,6 @@
-  - [Glossary](#glossary)
-
-<!-- /TOC -->
+<!-- TOC -->autoauto- [Vesting](#vesting)auto  - [Intro and Requirements](#intro-and-requirements)auto  - [Vesting Account Types](#vesting-account-types)auto  - [Vesting Account Specification](#vesting-account-specification)auto    - [Determining Vesting & Vested Amounts](#determining-vesting--vested-amounts)auto      - [Continuously Vesting Accounts](#continuously-vesting-accounts)auto      - [D ...
```
This looks really weird when just reading the file (e.g. in `vim`) - is this necessary for some reason?

fedekunze replied:
indeed if you display it in `rich diff` it looks weird as well

alexanderbez replied:
Woaaaaaa. I have zero idea what happened here or when? I will resolve this.

alexanderbez replied:
Fixed.

fedekunze commented on types/coin.go:
```diff
@@ -158,6 +158,23 @@ func (coins Coins) IsValid() bool {
 	}
 }
 
+// AddCoinByDenom adds a single coin to another coin in a set of coins by
```
lol what a tongue twister 😆

alexanderbez replied:
haha, fair. Updated.

fedekunze commented on types/coin.go:
```diff
@@ -226,6 +243,37 @@ func (coins Coins) safePlus(coinsB Coins) Coins {
 	}
 }
 
+// SubCoinByDenom subtracts a single coin from another coin in a set of coins by
```
ditto

alexanderbez replied:
Updated.

fedekunze commented on types/coin.go:
```diff
@@ -226,6 +243,37 @@ func (coins Coins) safePlus(coinsB Coins) Coins {
+
+	for i, coin := range coins {
+		if coin.Denom == other.Denom {
+			otherNeg := Coin{other.Denom, other.Amount.Neg()}
```
```suggestion
			otherNeg := sdk.NewCoin(other.Denom, other.Amount.Neg())
```

alexanderbez replied:
We're already in the coin package/implementation so I don't think there is a need to do this. See all the other methods as they also use the `Coin` literal.

alexanderbez replied:
Will resolve this convo, but reopen if you have any objections 😄

fedekunze commented on types/coin.go:
```diff
@@ -226,6 +243,37 @@ func (coins Coins) safePlus(coinsB Coins) Coins {
+			if coins[i].IsZero() {
+				// remove resulting zero coin
+				coins = append(coins[:i], coins[i+1:]...)
+			} else if !coins[i].IsNotNegative() {
```
hmmm not related for this scope but that function should be `IsNegative()`

alexanderbez replied:
Dude I know, but we're moving to unsigned integers anyway. If we do it, I agree that it should be in another PR.

fedekunze commented on x/bank/keeper.go:
```diff
@@ -246,3 +286,56 @@ func inputOutputCoins(ctx sdk.Context, am auth.AccountKeeper, inputs []Input, ou
+		}
+	}
+
+	return sdk.NewTags("action", []byte("delegateCoins")), nil
```
can we add the `accAddress` on the tags here? Otherwise it will only have the action tag which is not very useful alone... Also can we move the tags into a `tags/tags.go` file ?

alexanderbez replied:
Agreed. Updated.

fedekunze commented on x/bank/keeper.go:
```diff
@@ -246,3 +286,56 @@ func inputOutputCoins(ctx sdk.Context, am auth.AccountKeeper, inputs []Input, ou
+		}
+	}
+
+	return sdk.NewTags("action", []byte("undelegateCoins")), nil
```
ditto

alexanderbez replied:
Updated.

alexanderbez commented:
@cwgoes and @fedekunze addressed your comments.

I cleaned up the bank keeper a bit by updating the `Account` interface API. This allows us to not have to type check for vesting accounts.

rigelrozanski commented on types/coin.go:
```diff
@@ -226,6 +246,40 @@ func (coins Coins) safePlus(coinsB Coins) Coins {
+			if res[i].IsZero() {
+				// remove resulting zero coin
+				res = append(res[:i], res[i+1:]...)
+			} else if !res[i].IsNotNegative() {
```
this is weird because it will panic for "0" correct? we should probably replace with an `IsNegative()` function I imagine

alexanderbez replied:
Hmmm, if `res[i]` is 0 then it'll be removed (line #265) or am I missing/misunderstanding something?

cwgoes replied:
I think you're right but I also find the double negation and conditional-overlap confusing to read, why not just use `IsNegative()` ?

alexanderbez replied:
Ahh, that is because we do not have a `IsNegative`. I shall add this.

alexanderbez replied:
Updated.

cwgoes commented:
Needs a rebase.

alexanderbez commented:
@cwgoes updated.

cwgoes commented on types/coin.go:
```diff
@@ -158,6 +158,26 @@ func (coins Coins) IsValid() bool {
+		return Coins{other}
+	}
+
+	for i, coin := range res {
```
We could binary search here.

Maybe premature optimization though...

alexanderbez replied:
Agreed. I do think it's a premature optimization though. I have added a TODO to both methods and will create an issue for a `Coin` binary search on `Coins` 👍

cwgoes commented on types/coin.go:
```diff
@@ -226,6 +246,40 @@ func (coins Coins) safePlus(coinsB Coins) Coins {
+		return Coins{other}
+	}
+
+	for i, coin := range res {
```
Ditto on binary search, ditto on premature optimization.

alexanderbez replied:
See above.

cwgoes commented on x/auth/account.go:
```diff
@@ -121,12 +139,339 @@ func (acc *BaseAccount) SetSequence(seq uint64) error {
+			bva.DelegatedFree = bva.DelegatedFree.AddCoinByDenom(yCoin)
+		}
+
+		bva.Coins = bva.Coins.SubCoinByDenom(coin)
```
This is weird, we're modifying the account itself (passed by reference). Should the caller save it afterwards? We don't actually save it in the bank implementation, insofar as I can see...

cwgoes replied:
I guess we do set the returned coins, so it works. I still think this is confusing - either the function should modify the account and require that the account be written, or not modify the account and require an explicit `setCoins`, not do both.

cwgoes commented on x/auth/account.go:
```diff
@@ -121,12 +139,339 @@ func (acc *BaseAccount) SetSequence(seq uint64) error {
+			bva.DelegatedVesting = bva.DelegatedVesting.SubCoinByDenom(yCoin)
+		}
+
+		bva.Coins = bva.Coins.AddCoinByDenom(coin)
```
Ditto on the above modification / requiring `setCoins` comment.

cwgoes commented on x/auth/account.go:
```diff
@@ -121,12 +139,339 @@ func (acc *BaseAccount) SetSequence(seq uint64) error {
+
+// TrackDelegation performs delegation accounting. For a base account it simply
+// returns the base coins minus the desired delegation amount.
+func (acc *BaseAccount) TrackDelegation(blockTime time.Time, amount sdk.Coins) sdk.Coins {
```
Here we don't modify the account.

alexanderbez replied:
Updated.

cwgoes commented on x/auth/account.go:
```diff
@@ -121,12 +139,339 @@ func (acc *BaseAccount) SetSequence(seq uint64) error {
+// TrackUndelegation performs undelegation accounting. For a base account it
+// simply returns the base coins plus the undelegation amount.
+func (acc *BaseAccount) TrackUndelegation(amount sdk.Coins) sdk.Coins {
+	return acc.GetCoins().Plus(amount)
```
Here we don't modify the account (see below comments).

alexanderbez replied:
Updated.

cwgoes commented:
Tested ACK. Another final review or two, maybe by @rigelrozanski or @jaekwon, would be prudent.

alexanderbez commented:
@fedekunze mind giving another review?

alexanderbez commented on types/coin.go:
```diff
@@ -158,6 +158,28 @@ func (coins Coins) IsValid() bool {
+
+	// TODO: Perform binary search on coins to improve runtime performance
+	// (i.e. implement a generic binary coin search on Coins)
+	for i, coin := range res {
```
Done. There was already a test case for this, but I updated it to panic.

alexanderbez commented on types/coin.go:
```diff
@@ -226,6 +248,42 @@ func (coins Coins) safePlus(coinsB Coins) Coins {
+	}
+
+	return res
+}
```
Updated.

alexanderbez commented on types/coin.go:
```diff
@@ -226,6 +248,42 @@ func (coins Coins) safePlus(coinsB Coins) Coins {
+// SubCoinByDenom looks for a matching coin in coins that has the same denom as
+// other. Upon matching, the other coin is subtracted from the matching coin. If
+// the resulting coin is zero, it is removed. If the coins are empty, the coin
+// is simply added to the empty set.
```
Updated.

cwgoes commented:
Fails linter now :cry: .

alexanderbez commented:
Indeed! Will fix asap.

alexanderbez commented:
Indeed! Will fix asap.

cwgoes commented:
@alexanderbez Needs a minor rebase due to gas changes. Otherwise my approval still stands, but this needs another review.

alexanderbez commented:
Sigh...I'll update. This will keep happening until @jaekwon reviews imagine.

alexanderbez commented:
Rebased.

fedekunze commented on x/auth/account.go:
```diff
@@ -122,12 +140,335 @@ func (acc *BaseAccount) SetSequence(seq uint64) error {
+// returned.
+//
+// CONTRACT: The account's coins and undelegation coins must be sorted.
+func (bva *BaseVestingAccount) TrackUndelegation(amount sdk.Coins) {
```
ditto

fedekunze replied:
we're only checking if ≠ 0

alexanderbez replied:
I don't really think there is a need to? Are you asking to perform a `amount.IsValid()` call here? We know the coins cannot be negative and we already check the `0` case. So why incur the overhead?

fedekunze commented on x/auth/account.go:
```diff
@@ -122,12 +140,335 @@ func (acc *BaseAccount) SetSequence(seq uint64) error {
+//
+// CONTRACT: The account's coins, delegation coins, vesting coins, and delegated
+// vesting coins must be sorted.
+func (bva *BaseVestingAccount) trackDelegation(vestingCoins, amount sdk.Coins) {
```
can we check that the coins are > 0 ?

zmanian commented:
This is an extremely good candidate for the hard fork game of stakes upgrade after merged.

cwgoes commented:
I think this could be merged. Does anyone else want to review it?

jaekwon commented on docs/spec/auth/vesting.md:
```diff
@@ -74,28 +58,46 @@ type BaseVestingAccount struct {
+
 ## Vesting Account Specification
 
 Given a vesting account, we define the following in the proceeding operations:
```
In definition of BC, the sentence parsing is ambiguous or difficult to parse.  Use () instead of too many commas.

`The number of `OV` coins less any coins that are transferred (which can be negative) or delegated`

alexanderbez replied:
I fail to see how that is more clear than:

```
The number of `OV` coins less any coins that are transferred, which can be negative, or delegated (`DV + DF`)
```

Perhaps I can remove a comma?

alexanderbez replied:
Updated :)

jaekwon commented on types/coin.go:
```diff
@@ -172,6 +172,29 @@ func (coins Coins) IsValid() bool {
+
+	// TODO: Perform binary search on coins to improve runtime performance
+	// (i.e. implement a generic binary coin search on Coins)
+	for i, coin := range res {
```
AddCoinByDenom shouldn't panic when coin doesn't exist in coins.  The name of the function and the scope of functionality implies that it shouldn't.  Also it's not even true, when len(res) == 0, the comments are wrong.  This function shouldn't panic.

alexanderbez replied:
Hmmm you actually suggested the `SubCoinByDenom` variant SHOULD panic here: https://github.com/cosmos/cosmos-sdk/pull/2694#discussion_r239417664. Why not the `AddCoinByDenom` variant?

jaekwon replied:
Because there's no reason for this function to panic when implemented correctly.

alexanderbez replied:
Sure! So do you suggest we simply return `coins` when the denom isn't present? What's wrong with the implementation?

alexanderbez replied:
Updated :)

jaekwon commented on docs/spec/auth/vesting.md:
```diff
@@ -1,7 +1,5 @@
 # Vesting
```
Various grammatical errors throughout, needs a slow re-read.

alexanderbez replied:
Really? Can you please point them out? I've updated this spec several times as well as having read it many times.

jaekwon commented on docs/spec/auth/vesting.md:
````diff
@@ -74,28 +58,46 @@ type BaseVestingAccount struct {
 }
 ```
 
+In order to facilitate less ad-hoc type checking and assertions and to support
````
It would be better for x/bank to use runtime type checking to check for a vesting-account interface, than to require the auth Account interface to implement SpendableCoins/TrackDelegation/TrackUndelegation.  These methods should be declared in VestingAccount.  x/auth should know nothing of vesting, it should all be in x/bank.

If this isn't possible, please let me know why not.  If it is possible, we should probably do it, and not require x/auth accounts to know about vesting.

cwgoes replied:
I think it is possible - previously this was done (in some form) - but it made the bank code far more difficult to read and reason through, so we decided to abstract away this functionality in the account interface, which made it much easier to reason through.

Would moving `VestingAccount` to `x/bank` (but retaining the interface abstraction) help?

alexanderbez replied:
@cwgoes and I wanted to avoid type checking (as that was the way it was initially) as that pollutes the code and makes it less flexible.

>  x/auth should know nothing of vesting, it should all be in x/bank.

It would technically, as it would have to type check.

> If this isn't possible, please let me know why not. If it is possible, we should probably do it, and not require x/auth accounts to know about vesting.

It is possible and that was the way it was initially intended, but there was a bit of dup code and was not as clean. I do not want to keep going back-n-forth on this and just simply decide as this PR has been open for a long time now.

alexanderbez replied:
I vaguely recall us having a phone convo about this 🤔 Wouldn't it also be weird to have a `VestingAccount` in `x/bank` as bank should not really define account types?

jaekwon commented on docs/spec/auth/vesting.md:
```diff
@@ -74,28 +58,46 @@ type BaseVestingAccount struct {
+
+    // Delegation and undelegation accounting that returns the resulting base
+    // coins amount.
+    TrackDelegation(Time, Coins)
```
TrackDelegation/TrackUndelegation are weird names.  DelegateCoins/UndelegateCoins seems fine?

alexanderbez replied:
Ehhh, but you're not actually delegating or undelegating. You're simply performing some accounting. I think `Track*` is just fine imho.

sunnya97 commented:
If we don't prioritize Undelegating into `VestedCoins`, couldn't we get rid of tracking of `DelegatedVested` vs `DelegatedVesting`?  Wouldn't that simplify things a lot?

Furthermore, can we change the name of `bankKeeper.DelegateCoins()` to `bankKeeper.SubtractVestin()` or something, cause that's really what the desired functionality of this feature is.  For example, you should be able to use VestingCoins for governance deposits.

sunnya97 commented:
Essentially I want to reduce the breaking of the abstraction barriers here between `stake` and `bank`.  I don't like that right now `bank` knows about staking.  Currently the dependency is that `stake` knows about `bank`, not the other way around.

> So you'd like to call it SubtractVesting? What if the account isn't a vesting account? I'm not in favor of this renaming.

Hmm, you're right.  I don't like that either.  An alternative:  A `bank.VestedOnlyKeeper,` that still meets the `bank.Keeper` interface definition, but caps at `VestedCoins` for SubtractCoins.  So keepers that should have access to `Vesting` coins get the normal Keeper, while ones that don't get the VestedOnlyKeeper.  The converse of this is also fine (`VestingAccessibleKeeper` & `NormalKeeper`).

> But why not support undelegating? It works.

It gets rid of the tracking of delegated coins, simplifying this module significantly.  Is there a reason we care to prioritize vested coins?  (this only matters if coins are slashed/burned right?)

> I think we talked about that and we decided not to support such a feature.

Oh, I didn't realize that.  What was the reason for this?  Was this discussion recorded somewhere?

alexanderbez commented:
Ok @jaekwon and @cwgoes, I've updated the code and spec to address the major comments, namely:

- `AddCoinByDenom` no longer panics
- Removed `TrackDelegation` and `TrackUndelegation` from the `Account` interface. The `x/bank` keeper now simply calls private auxiliary functions: `trackDelegation` and `trackUndelegation` which do type checking.
- Updated spec and fixed a few typos

At this point, I think the vesting spec and implementation are in a good place -- lmk your thoughts.

cwgoes commented:
@alexanderbez Thanks; your changes look fine to me - my approval stands.

jackzampolin commented on types/coin.go:
```diff
@@ -172,6 +172,30 @@ func (coins Coins) IsValid() bool {
+// coins are empty, the coin is simply added to the empty set. If the denom of
+// the other coin to be added does not exist in coins, the function returns the
+// original coins.
+func (coins Coins) AddCoinByDenom(other Coin) Coins {
```
So this function will return `Coins{Coin}` if `Coins` is empty, but if `Coins` doesn't contain any of `Coin.denom` then it doesn't return with `Coin` included in `Coins`. That seems like it violates expectations. Can we make it so `Coins` includes `Coin` if the denom isn't present in `Coins`? cc @alexanderbez

alexanderbez replied:
I'll revise it to not add `coin` when the `coins` set is empty.

cwgoes commented:
@jackzampolin Can you check if your suggestions have been addressed?

@jaekwon Did you have any further suggestions or concerns?

alexanderbez commented:
@cwgoes I spoke with @jaekwon -- I'll be removing `Add/SubCoinsByDenom` in favor of the original `Plus/Sub` logic. Afterwards, it can be merged.
